In [1]:
import pandas as pd

In [2]:
try:
    df = pd.read_csv(r'C:\Users\Иванов\Downloads\data.csv')
except:
    df = pd.read_csv('/datasets/data.csv')

In [3]:
display(df.head(10))

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

* children — количество детей в семье
* days_employed — общий трудовой стаж в днях
* dob_years — возраст клиента в годах
* education — уровень образования клиента
* education_id — идентификатор уровня образования
* family_status — семейное положение
* family_status_id — идентификатор семейного положения
* gender — пол клиента
* income_type — тип занятости
* debt — имел ли задолженность по возврату кредитов
* total_income — ежемесячный доход
* purpose — цель получения кредита

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### В двух столбцах есть пропущенные значения, найдите их и заполните медианным значением по столбцу:
- опишите, какие пропущенные значения вы обнаружили;
- проверьте, какую долю составляют пропущенные значения в каждом из столбцов с пропусками;
- приведите возможные причины появления пропусков в данных;
- объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. Например, отрицательное количество дней трудового стажа в столбце days_employed. Для реальных данных это нормально. Обработайте значения в столбцах с аномалиями и опишите возможные причины появления таких данных.

Замените вещественный тип данных в столбце total_income на целочисленный, например, с помощью метода astype().

Если в данных присутствуют строки-дубликаты, удалите их. Также обработайте неявные дубликаты. Например, в столбце education есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к одному регистру. Проверьте остальные столбцы. 

После удаления дубликатов сделайте следующее:
поясните, как выбирали метод для поиска и удаления дубликатов в данных;
приведите возможные причины появления дубликатов.

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Столбцы days_employed (общий трудовой стаж в днях) и total_income (ежемесячный доход) имеют пропущенные значения.</div>


In [5]:
df[df['days_employed'].isna()].head(20)

children  days_employed  dob_years education  education_id  \
12          0            NaN         65   среднее             1   
26          0            NaN         41   среднее             1   
29          0            NaN         63   среднее             1   
41          0            NaN         50   среднее             1   
55          0            NaN         54   среднее             1   
65          0            NaN         21   среднее             1   
67          0            NaN         52    высшее             0   
72          1            NaN         32    высшее             0   
82          2            NaN         50    высшее             0   
83          0            NaN         52   среднее             1   
90          2            NaN         35    высшее             0   
94          1            NaN         34    высшее             0   
96          0            NaN         44   СРЕДНЕЕ             1   
97          0            NaN         47    высшее             0   
120         0            NaN         46   среднее             1   
121         0            NaN         29    высшее             0   
135         0            NaN         27   среднее             1   
141         0            NaN         39   среднее             1   
145         0            NaN         62   среднее             1   
174         0            NaN         55    высшее             0   

             family_status  family_status_id gender  income_type  debt  \
12        гражданский брак                 1      M    пенсионер     0   
26         женат / замужем                 0      M  госслужащий     0   
29   Не женат / не замужем                 4      F    пенсионер     0   
41         женат / замужем                 0      F  госслужащий     0   
55        гражданский брак                 1      F    пенсионер     1   
65   Не женат / не замужем                 4      M    компаньон     0   
67         женат / замужем                 0      F    пенсионер     0   
72         женат / замужем                 0      M  госслужащий     0   
82         женат / замужем                 0      F    сотрудник     0   
83         женат / замужем                 0      M    сотрудник     0   
90         женат / замужем                 0      F    сотрудник     0   
94        гражданский брак                 1      F    компаньон     0   
96         женат / замужем                 0      F    сотрудник     0   
97         женат / замужем                 0      F    сотрудник     0   
120        женат / замужем                 0      F    сотрудник     0   
121        женат / замужем                 0      F    компаньон     0   
135        женат / замужем                 0      M    компаньон     0   
141       гражданский брак                 1      M    сотрудник     0   
145        женат / замужем                 0      M    пенсионер     0   
174         вдовец / вдова                 2      F    компаньон     0   

     total_income                                purpose  
12            NaN                        сыграть свадьбу  
26            NaN                            образование  
29            NaN       строительство жилой недвижимости  
41            NaN       сделка с подержанным автомобилем  
55            NaN                        сыграть свадьбу  
65            NaN  операции с коммерческой недвижимостью  
67            NaN                покупка жилья для семьи  
72            NaN  операции с коммерческой недвижимостью  
82            NaN                                  жилье  
83            NaN                                  жилье  
90            NaN                      операции с жильем  
94            NaN                  на проведение свадьбы  
96            NaN             покупка жилой недвижимости  
97            NaN                 профильное образование  
120           NaN                     высшее образование  
121           NaN                             автомобиль  
135           NaN                      

In [6]:
df_nan = df[df['days_employed'].isna()]

In [7]:
df[df['days_employed'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          2174 non-null   int64  
 1   days_employed     0 non-null      float64
 2   dob_years         2174 non-null   int64  
 3   education         2174 non-null   object 
 4   education_id      2174 non-null   int64  
 5   family_status     2174 non-null   object 
 6   family_status_id  2174 non-null   int64  
 7   gender            2174 non-null   object 
 8   income_type       2174 non-null   object 
 9   debt              2174 non-null   int64  
 10  total_income      0 non-null      float64
 11  purpose           2174 non-null   object 
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB


<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Пропуски в данных у клиентов и в столбе 'days_employed' и в 'total_income'. 

Причиной этого может быть нежелание клиента раскрывать свой ежемесячный доход, а программа для заполнения данных автоматически считает их неработающими. В итоге в столбце 'days_employed' проставляя пропуски.

Вряд ли таким клиентам выдадут кредит...</div>

In [8]:
len(df[df['days_employed'].isna()])

2174

In [9]:
len(df[df['total_income'].isna()])

2174

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Считаем сколько пропусков в каждом из столбцов. В обоих 2174.</div>

In [10]:
len(df[df['days_employed'].isna()])/len(df['days_employed'])*100

10.099883855981417

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Чуть больше 10%</div>

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Среднее значение использовать некорректно, так как некоторые значения могут сильно выделяться среди большинства. Поэтому будем использовать медианное значение. 
    
    НО!
По заданию проекта первым действием необходимо заполнить пропуски в столбцах, а затем искать и исправлять аномалии. Считаю это немного не корректным.

Объясню свою позицию:

столбец days_employed (общий трудовой стаж в днях) имеет отрицательные значения. Очевидно это аномалия. Вызванна, вероятно, сбоем программы или человеческим фактором. Для исправления этого необходимо переписать эти значения с положительным знаком, т.е. по модулю. В таком случае и медианные значения столбца изменятся  2194.220566878695 против -1203.369528770489.

Итак приступаю к поиску и исправлению аномалий и дубликатов.</div>

In [11]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

In [12]:
print(df['children'].value_counts())

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64


<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Столбец children — количество детей в семье. 

У кого-то количество детей "-1". Исправляем модулем!

А у кого-то "20"! У 76 человек... При этом количество детей от 6 до 19 отсутствует. Исправляем на медианное!</div>

In [13]:
df['children'] = df['children'].abs()

In [14]:
children_median = df['children'].median()
df.loc[df['children']==20, 'children'] = children_median

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Столбец days_employed — общий трудовой стаж в днях.

1. Отрицательные значения. Исправляем!</div>

In [15]:
df['days_employed'] = df['days_employed'].abs()

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    2. 401755 часов... Это 401755/365. Около 1100 лет работы!!!

Согласно даннм ПФР [ссылка тут](https://ria.ru/20131119/977986173.html) средний трудовой стаж 34.5 года, т.е. 12592 часа

Допустим, что человек за жизнь мог набрать максимально за 75 лет жизни (максимальный возраст в таблице) работая с 16 лет и всю жизнь ("Богатыри - не Вы!"(С)). 59 лет = 21535 часов.

Проблема в том, что такой человек в таблице у нас всего один...</div>

In [16]:
df_work = df[df['days_employed']>21535]

In [17]:
display(df_work.groupby('income_type')['education'].count())

income_type
безработный       2
пенсионер      3443
Name: education, dtype: int64

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Основной тип занятости у людей с превышенным значением общего трудового стажа - "пенсонер"! И таких 3443 человека из 21525.
Исправляем на значение среднего трудового стажа по данным ПФР т.е. на 12592 часа.</div>

In [18]:
df.loc[df['days_employed']>12592, 'days_employed'] = 12592

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Теперь можно заполнить пропуски медианным значением.</div>

In [19]:
days_employed_median = df['days_employed'].median()

In [20]:
df['days_employed'] = df['days_employed'].fillna(days_employed_median)

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Столбец dob_years — возраст клиента в годах. Кто-то хочет получить кредит ещё не родясь... </div>

In [21]:
print(df['dob_years'].value_counts())

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
22    183
66    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64


<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    101 человек. Надо бы сделать это поле обязательным для заполнения в программе! Мы изменяем значения на медианные.</div>

In [22]:
dob_years_median = df['dob_years'].median()

In [23]:
df.loc[df['dob_years']==0, 'dob_years'] = dob_years_median

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Столбец education — уровень образования клиента. Приводим все значения в нижний регистр.</div>

In [24]:
df['education'] = df['education'].str.lower()

In [25]:
print(df['education'].value_counts())

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64


<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Столбец education_id — идентификатор уровня образования.</div>

In [26]:
print(df['education_id'].value_counts())

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64


<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Оставляем.</div>

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Столбец family_status — семейное положение. Приводим значения в нижний регистр.</div>

In [27]:
df['family_status'] = df['family_status'].str.lower()

In [28]:
print(df['family_status'].value_counts())

женат / замужем          12380
гражданский брак          4177
не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64


<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Столбец family_status_id — идентификатор семейного положения. Смотрим. Идем дальше.</div>

In [29]:
print(df['family_status_id'].value_counts())

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64


<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Столбец gender — пол клиента.</div>

In [30]:
print(df['gender'].value_counts())

F      14236
M       7288
XNA        1
Name: gender, dtype: int64


<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Кто такой/такая XNA??? </div>

In [31]:
display(df[df['gender'] == 'XNA'])

children  days_employed  dob_years            education  education_id  \
10701         0    2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Исправляем.</div>

In [32]:
df.loc[df['gender'] == 'XNA', 'gender'] = 'F'

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Столбец income_type — тип занятости. Проверяем.</div>

In [33]:
print(df['income_type'].value_counts())

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64


<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Пропускаем.

Столбец debt — имел ли задолженность по возврату кредитов. Проверяем.</div>

In [34]:
print(df['debt'].value_counts())

0    19784
1     1741
Name: debt, dtype: int64


<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Пропускаем. Столбец total_income — ежемесячный доход. Переводим в целоцисленный тип данных. Заполняем пропуски медианным значением.</div>

In [35]:
total_income_median = df['total_income'].median()
df['total_income'] = df['total_income'].fillna(total_income_median)

In [36]:
df['total_income'] = df['total_income'].astype('int')

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Крайний столбец purpose — цель получения кредита. Смотрим.</div>

In [37]:
print(df['purpose'].value_counts())

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Проверку данных на наличие явных дубликатов проводить нет смысла, т.к. клиенты не имеют уникального номера, соответственно даже если данные таблицы идентичны, это не значит что это один и тот же человек. Неявные дубликаты найдены и исправлены выше. Возможные причины появления дубликатов - человеческий фактор при работе с данными и некорректная работа программы.</div>

### Создайте два новых датафрейма со столбцами:

education_id и education — в первом;

family_status_id и family_status — во втором.

Удалите из исходного датафрейма столбцы education и family_status, оставив только их идентификаторы: education_id и family_status_id. 

Новые датафреймы — это те самые «словари» (не путайте с одноимённой структурой данных в Python), к которым вы сможете обращаться по идентификатору.

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Создаем датафреймы, удаляем дубликаты, перезаписываем индексы.</div>

In [38]:
education_log = df[['education_id', 'education']].drop_duplicates().reset_index(drop=True)
display(education_log)

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

In [39]:
family_status_log = df[['family_status_id', 'family_status']].drop_duplicates().reset_index(drop=True)
display(family_status_log)

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Удаляем из исходного датафрейма столбцы education и family_status</div>

In [40]:
df.drop(columns=['education', 'family_status'], axis=1, inplace=True)

In [41]:
display(df.head())

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0   12592.000000         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

### На основании диапазонов, указанных ниже, создайте столбец total_income_category с категориями:
* 0–30000 — 'E';
* 30001–50000 — 'D';
* 50001–200000 — 'C';
* 200001–1000000 — 'B';
* 1000001 и выше — 'A'.

Например, кредитополучателю с доходом 25000 нужно назначить категорию 'E', а клиенту, получающему 235000, — 'B'.

In [42]:
def income_group(income):
    if 0 <= income <= 30000:
        return 'E'
    if 30001 <= income <= 50000:
        return 'D'
    if 50001 <= income <= 200000:
        return 'C'
    if 200001 <= income <= 1000000:
        return 'B'
    return 'A'
df['total_income_category'] = df['total_income'].apply(income_group)

In [43]:
display(df.head())

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0   12592.000000         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

### Создайте функцию, которая на основании данных из столбца purpose сформирует новый столбец purpose_category, в который войдут следующие категории:
* ‘операции с автомобилем’,
* ‘операции с недвижимостью’,
* ‘проведение свадьбы’,
* ‘получение образования’.

Например, если в столбце purpose находится подстрока 'на покупку автомобиля', то в столбце purpose_category должна появиться строка 'операции с автомобилем'.
Вы можете использовать собственную функцию и метод apply(). Изучите данные в столбце purpose и определите, какие подстроки помогут вам правильно определить категорию.

In [44]:
print(df['purpose'].value_counts())

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

In [45]:
def purpose_group(purpose):
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    if 'авто' in purpose:
        return 'операции с автомобилем'
    if 'жил' in purpose or 'недвижимос' in purpose:
        return 'операции с недвижимостью'
    if 'образова' in purpose:
        return 'получение образования'
    return 'A'
    
df['purpose_category'] = df['purpose'].apply(purpose_group)

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    return 'A' сделано для того, что бы увидеть что все данные из столбца 'purpose' охвачены и обработаны.</div>

In [46]:
print(df['purpose_category'].value_counts())

операции с недвижимостью    10840
операции с автомобилем       4315
получение образования        4022
проведение свадьбы           2348
Name: purpose_category, dtype: int64


In [47]:
display(df.head(5))

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0   12592.000000         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

### Шаг 3. Ответьте на вопросы
Ответы на вопросы можно разместить в ячейках тетрадок Jupyter Notebook с типом markdown.
* Есть ли зависимость между количеством детей и возвратом кредита в срок?
* Есть ли зависимость между семейным положением и возвратом кредита в срок?
* Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
* Как разные цели кредита влияют на его возврат в срок?

Ответы сопроводите интерпретацией — поясните, о чём именно говорит полученный вами результат.

In [48]:
children_debt = df.pivot_table(index=['children'], columns='debt', values='dob_years', aggfunc='count')
children_debt['total'] = (children_debt[0]+children_debt[1])
children_debt['percent'] = (children_debt[1]/children_debt['total'])*100
display(children_debt)

debt            0       1    total   percent
children                                    
0         13154.0  1071.0  14225.0  7.528998
1          4420.0   445.0   4865.0  9.146968
2          1861.0   194.0   2055.0  9.440389
3           303.0    27.0    330.0  8.181818
4            37.0     4.0     41.0  9.756098
5             9.0     NaN      NaN       NaN

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Зависимость между количеством детей и возвратом кредита в срок прослеживается. Как видно из таблицы, процент должников по кредиту растёт с увеличением количества детей. За исключением если детей трое. По моему субъективному мнению статистикой наличия детей от 3+ можно пренебречь, т.к. таких клиентов 371 из 21525 (1,7%). А вот статистика количества детей от 0 до 2 влючительно явно даёт понять что чем больше детей, тем сложнее выплачивать кредит.</div>

In [49]:
family_debt = df.pivot_table(index=['family_status_id'], columns='debt', values='dob_years', aggfunc='count')
family_debt['family_status'] = family_status_log['family_status']
family_debt['total'] = (family_debt[0]+family_debt[1])
family_debt['percent'] = (family_debt[1]/family_debt['total'])*100
display(family_debt)

debt                  0    1          family_status  total   percent
family_status_id                                                    
0                 11449  931        женат / замужем  12380  7.520194
1                  3789  388       гражданский брак   4177  9.288963
2                   897   63         вдовец / вдова    960  6.562500
3                  1110   85              в разводе   1195  7.112971
4                  2539  274  не женат / не замужем   2813  9.740491

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Из таблицы видно, что самые безответственные имеют статус "не женат / не замужем" и "гражданский брак"

Кстати, понятие "гражданский брак" не совсем соответствует тому, чем сейчас это называют. 

[если интересно - ссылка тут](https://ru.wikipedia.org/wiki/%D0%93%D1%80%D0%B0%D0%B6%D0%B4%D0%B0%D0%BD%D1%81%D0%BA%D0%B8%D0%B9_%D0%B1%D1%80%D0%B0%D0%BA#:~:text=%D0%93%D1%80%D0%B0%D0%B6%D0%B4%D0%B0%CC%81%D0%BD%D1%81%D0%BA%D0%B8%D0%B9%2C%20%D0%B8%D0%BB%D0%B8%20%D1%81%D0%B2%D0%B5%CC%81%D1%82%D1%81%D0%BA%D0%B8%D0%B9%2C%20%D0%B1%D1%80%D0%B0%D0%BA%20%E2%80%94,%D0%B8%D1%81%D0%BF%D0%BE%D0%BB%D1%8C%D0%B7%D0%BE%D0%B2%D0%B0%D1%82%D1%8C%D1%81%D1%8F%20%D0%B4%D0%BB%D1%8F%20%D0%BE%D0%B1%D0%BE%D0%B7%D0%BD%D0%B0%D1%87%D0%B5%D0%BD%D0%B8%D1%8F%20%D1%84%D0%B0%D0%BA%D1%82%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%BE%D0%B3%D0%BE%20%D1%81%D0%BE%D0%B6%D0%B8%D1%82%D0%B5%D0%BB%D1%8C%D1%81%D1%82%D0%B2%D0%B0.)

А самыми ответственными являются вдовцы/вдовы.</div>

In [50]:
income_debt = df.pivot_table(index=['total_income_category'], columns='debt', values='dob_years', aggfunc='count')
income_debt['total'] = (income_debt[0]+income_debt[1])
income_debt['percent'] = (income_debt[1]/income_debt['total'])*100
display(income_debt)

debt                       0     1  total   percent
total_income_category                              
A                         23     2     25  8.000000
B                       4685   356   5041  7.062091
C                      14727  1360  16087  8.454031
D                        329    21    350  6.000000
E                         20     2     22  9.090909

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Самый неплатёжеспособный класс с ежемесячным доходом до 30000 р. Логично. Даже прожить на эти деньги сложно. 
Хотя для статистических выводов количество в 22 человека слишком мало.

Если учитывать самые большие классы по количесву клиентов (В и С) 
зависимость платежеспособности пропорциональна ежемесячному доходу.</div>

In [51]:
purpose_debt = df.pivot_table(index=['purpose_category'], columns='debt', values='dob_years', aggfunc='count')
purpose_debt['total'] = (purpose_debt[0]+purpose_debt[1])
purpose_debt['percent'] = (purpose_debt[1]/purpose_debt['total'])*100
display(purpose_debt)

debt                          0    1  total   percent
purpose_category                                     
операции с автомобилем     3912  403   4315  9.339513
операции с недвижимостью  10058  782  10840  7.214022
получение образования      3652  370   4022  9.199403
проведение свадьбы         2162  186   2348  7.921635

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Самые платёжеспособные клиенты те, кто желает произвести какие-либо операции с недвижимостью. Возможно, это связано с залогом недвижимости, что является несомненным стимулом при расчётах с банком. Не отстают и новобрачные (новая семья - новая жизнь без долгов). А вот автомобилисты и желающие обучаться отстают. Видимо, связано с последующем разочарованием от покупки (как хорошо, что среди брачующихся нет такой тенденции).</div>

### Шаг 4. Напишите общий вывод
Оформление: Задание выполните в Jupyter Notebook. Программный код заполните в ячейках типа code, текстовые пояснения — в ячейках типа markdown. Примените форматирование и заголовки.

<div class="alert alert-info" style="box-shadow: 4px 4px 4px">
    Из всего изложенного можно сделать выводы:
    
* Дети - цветы жизни! Но, на них нужно много денег. Многодетные менее платёжеспособны, но это - не повод не давать им кредиты.
* Неженатые/незамужние и "гражданско брачующие" менее ревностно относятся к кредитным обязательствам.
* Меньше зарабатываешь - меньше возможности платить кредит. Логично.
* Клиент хочет недвижимость или бракосочетаться? Святое дело оформить кредит!

Уважаемый ревьютер! Прошу, не ругаться за свободное описание моих выводов и отступлений от темы. Проект делал в то время, когда предыдущий ещё не проверили, формы проекта у меня ещё не было. Буду рад всем твоим комментариям, особенно плохим! (это поможет мне стать лучше). Удачи!</div>
